packages/mdc-toolbar/foundation.js

```diff
@@ -0,0 +1,234 @@
+    this.adapter_.registerScrollHandler(this.scrollHandler_);
+  }
+
+  destroy() {
+    this.adapter_.deregisterResizeHandler(this.resizeHandler_);
+    this.adapter_.deregisterScrollHandler(this.scrollHandler_);
+  }
+
+  setKeyHeights() {
+    if (this.resizeFrame_ !== 0) {
+      cancelAnimationFrame(this.resizeFrame_);
+    }
+    this.resizeFrame_ = requestAnimationFrame(() => {
+      this.setKeyHeights_();
+    });
+  }
+
+  changeAdjustElementStyles() {
+    if (this.fixedLastrow_ || this.fixedAllRow_) {
+      this.adapter_.setStyleForFixedAdjustElement('marginTop', `${this.keyHeights.toolbarHeight}px`);
+    }
+  }
+
+  changeToolbarStyles() {
+    if (this.scrollFrame_ !== 0) {
+      cancelAnimationFrame(this.scrollFrame_);
+    }
+
+    this.scrollFrame_ = requestAnimationFrame(() => {
+      const scrollTop = this.adapter_.getViewportScrollY();
+
+      if (this.scrolledOutOfTheshold_(scrollTop)) {
+        return; // return early if it is already out of theshold.
+      }
+
+      const flexibleExpansionRatio = this.getFlexibleExpansionRatio(scrollTop);
+      this.adapter_.notifyChange(flexibleExpansionRatio);
+      this.changeToolbarFlexibleState_(flexibleExpansionRatio);
+      if (this.fixedLastrow_) {
+        this.changeToolbarFixedState_(scrollTop);
+      }
+      if (this.hasFlexibleRow_) {
+        this.changeFlexibleRowElementStyles_(flexibleExpansionRatio);
+      }
+    });
+  }
+
+  getFlexibleExpansionRatio(scrollTop) {
+    const delta = 0.0001;
+    return Math.max(0, 1 - scrollTop / (this.keyHeights.flexibleExpansionHeight + delta));
+  }
+
+  initKeyRatio_() {
+    this.keyHeights.toolbarRowHeight = this.getRowHeight_();
+    const flexibleRowMaxRatio = this.adapter_.getFlexibleRowElementOffsetHeight() / this.keyHeights.toolbarRowHeight;
+    this.keyHeights.toolbarRatio = this.adapter_.getOffsetHeight() / this.keyHeights.toolbarRowHeight;
+    this.keyHeights.flexibleExpansionRatio = flexibleRowMaxRatio - 1;
+    this.keyHeights.maxTranslateYRatio = this.fixedLastrow_? this.keyHeights.toolbarRatio - flexibleRowMaxRatio : 0;
+    this.keyHeights.scrollThesholdRatio = (this.fixedLastrow_? this.keyHeights.toolbarRatio : flexibleRowMaxRatio) - 1;
+  }
+
+  getRowHeight_() {
+    const breakpoint = MDCToolbarFoundation.numbers.TOOLBAR_MOBILE_BREAKPOINT;
+    return this.adapter_.getViewportWidth() <= breakpoint ?
+      MDCToolbarFoundation.numbers.TOOLBAR_ROW_MOBILE_HEIGHT : MDCToolbarFoundation.numbers.TOOLBAR_ROW_HEIGHT;
+  }
+
+  scrolledOutOfTheshold_(scrollTop) {
+    if (scrollTop > this.keyHeights.scrollTheshold) {
+      if (this.executedLastChange_) {
+        return true;
+      }
+      this.executedLastChange_ = true;
+    } else {
+      this.executedLastChange_ = false;
+    }
+    return false;
+  }
+
+  setKeyHeights_({isInit = false} = {}) {
+    const newToolbarRowHeight = this.getRowHeight_();
+    if (newToolbarRowHeight !== this.keyHeights.toolbarRowHeight || isInit) {
```

Comment:
It was implemented this way because I was setting `this.keyHeights.toolbarRowHeight` in `initKeyRatio_` which essentially run before `setKeyHeights_` is run. However, you remind me that I totally don't need rush to set that value in `initKeyRatio_` but could use a local constant and defer the set of `this.keyHeights.toolbarRowHeight` to `setKeyHeights_`, which seems to me make things much cleaner and remove the need for this `isInit` flag. See if the new implementation feels better to you.